tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -5889,6 +5896,33 @@ bool HLSLExternalSource::MatchArguments(
   DXASSERT(isVariadic || iterArg == end,
            "otherwise the argument list wasn't fully processed");
 
+  // Second pass replaces literal types when not supported for constant eval,
+  // and they haven't been resolved by other arguments.
+  for (auto arg : LiteralArgsToConvert) {
+    size_t iArg = arg.first;
+    Expr* pCallArg = arg.second;
+    const HLSL_INTRINSIC_ARGUMENT *pIntrinsicArg = &pIntrinsic->pArgs[iArg];
+
+    QualType pType = pCallArg->getType();
+    ArBasicKind TypeInfoEltKind = GetTypeElementKind(pType);
+
+    TypeInfoEltKind = ConcreteLiteralType(
+        pCallArg, TypeInfoEltKind, pIntrinsicArg->uLegalComponentTypes, this);
+
+    // Merge ComponentTypes
+    if (AR_BASIC_UNKNOWN == ComponentType[pIntrinsicArg->uComponentTypeId]) {
+      ComponentType[pIntrinsicArg->uComponentTypeId] = TypeInfoEltKind;
+    }
+    else {
+      if (!CombineBasicTypes(
+            ComponentType[pIntrinsicArg->uComponentTypeId],
+            TypeInfoEltKind,
+            &ComponentType[pIntrinsicArg->uComponentTypeId])) {
```

Comment:
Ideally, I think in the type-combine case, we would want to preserve the original literal when combining types here, only converting the literal to a concrete type when no type was ever assigned to this slot after the first pass (the AR_BASIC_UNKNOWN case).  I'll update the PR with this change.